README.md
<p align="center">
<img src="images/transformer.png"/>
<br>
<a href="https://travis-ci.org/zalando-incubator/Transformer"><img src="https://travis-ci.org/zalando-incubator/Transformer.svg?branch=master"/></a>
<a href="https://www.codacy.com/app/thilp/Transformer"><img src="https://api.codacy.com/project/badge/Grade/10b3feb4e4814429bf288b87443a6c72"/></a>
<a href="https://www.codacy.com/app/thilp/Transformer"><img src="https://api.codacy.com/project/badge/Coverage/10b3feb4e4814429bf288b87443a6c72"/></a>
</p>

# Transformer

A tool to convert web browser sessions ([HAR files][]) into
[Locust][] load testing scenarios (locustfiles).

Use it when you want to **replay HAR files** (containing recordings of
interactions with your website) **in load tests with Locust**.

[HAR files]: https://en.wikipedia.org/wiki/.har
[Locust]: https://locust.io/

<!-- toc -->

- [Installation](#installation)
- [Usage](#usage)
  * [Command-line](#command-line)
  * [Library](#library)
- [Documentation](#documentation)
- [Authors](#authors)
- [License](#license)

<!-- tocstop -->

## Installation

Install from PyPI with pip:

```bash
pip install har-transformer
```

## Usage

### Command-line

```bash
$ transformer my_har_files_dir/
```

### Library

```python
import transformer

with open("locustfile.py", "w") as f:
    transformer.dump(f, ["my_har_files_dir/"])
```

Example HAR files are included in the `examples` directory for you to try out.

## Documentation

Take a look at our **[user documentation][doc]** for more details, including
how to generate HAR files, customize your scenarios, use or write plugins …

[doc]: https://github.com/zalando-incubator/transformer/wiki

## Authors

* **Serhii Cherniavskyi** - [@scherniavsky](https://github.com/scherniavsky)
* **Thibaut Le Page** - [@thilp](https://github.com/thilp)
* **Brian Maher** - [@bmaher](https://github.com/bmaher)
* **Oliwia Zaremba** - [@tortila](https://github.com/tortila)

See also the list of [contributors](CONTRIBUTORS.md) who participated in this project.

## License

This project is licensed under the MIT License - see the
[LICENSE.md](LICENSE.md) file for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased][]

### Added

  - `transformer.request.Request.name`: Controls
  [Locust's URL grouping][locust-dynamic-parameters]. Its default value is
  `Request.url`, which ensures retrocompatibility. (#32)
  - `transformer.python.ExpressionView`: An Expression that wraps a non-Expression
  (e.g. a Request instance), similarly to how Standalone is a Statement that
  wraps an Expression. A ExpressionView has a `target` (the wrapped object), a
  `converter` (function capable of transforming the target into an Expression),
  and a `name` for inspection purposes. (#33)
  - A functional test suite in `functional-tests/`. It is part of `make test`
  and can be run alone with `make functest`. (#37)

### Changed

  - When processing HAR requests with the `application/json` MIME type,
  Transformer no longer uses the `params` field as a replacement for a missing
  `text` field. This was a mitigation for a bug in a different, internal tool. (#33)

### Fixed

  - A bug in the conversion between Task and Task2 makes Transformer ignore all
  changes made by plugins to `Task2.request` in the generated locustfile.
  Thank you [@xinke2411][] for reporting this! (#33)

### Removed

  - `transformer.task.Task.as_locust_action`: As part of the merge between Task
  and Task2 (#11). `as_locust_action` generates locustfile code as a string,
  which is made obsolete by the `transformer.python` syntax tree framework. (#33)
  - `transformer.decision`: Unused API for combining a boolean (the decision
  itself) with a string (a user-friendly reason for that decision). (#36)

## [1.0.2][] - 2019-02-21

### Added

  - `transformer.transform.{dump,dumps}`: Similar to `json.{dump,dumps}`, these
  high-level functions should be all most users need to know about Transformer.
  They convert lists of scenario paths and plugins into a final locustfile.
  They will replace the `transform` function, which requires more familiarity
  with Transformer's internals. (#14)
  - `transformer.locust.locustfile_lines`: Similar to `locustfile` but returns
  an iterator over lines (as strings) instead of a string containing the full
  locustfile. This design allows for more flexibility in `dump`/`dumps` and
  should result in smaller memory usage for huge locustfiles. (#14)
  - Preliminary support for new-generation plugins. (#25)

### Changed

  - The `dump`, `dumps`, and `transform` functions by default use the
  `sanitize_headers` plugin even if users don't provide it in the plugin list.
  This is because the resulting locustfile would almost certainly be broken
  without these plugins. Users can still opt-out from these default plugins
  by passing the keyword-argument `with_default_plugins=False` (e.g. if they
  implemented their own version). (#14)

### Deprecated

  - `transformer.transform.transform` (#14)
  - `transformer.locust.locustfile` (#14)

## [1.0.1][] - 2019-02-12

### Fixed

  - Fix `transformer` command-line crash due to a missing version identifier. (#17)
  - Publish development releases to PyPI for every merge to the `master` branch. (#17)

## [1.0.0][] - 2019-02-11

### Added

  - Plugin contracts aiming to eventually deprecate OnTaskSequence:
  OnTask, OnScenario, OnPythonProgram.
  - Transformer can now be called using a `transformer` script installed by
  `pip`, or via `python -m transformer`. (#7)
  - Transformer is now published to PyPI as [`har-transformer`] (it looks like
  the name `transformer` is already taken, unsurprisingly). (#3)
  - Remove Pipenv completely; everything uses [Poetry] now.
  - Add a general `transformer.plugins.contracts.Plugin` type covering all more
  specialized plugin types, like `OnTask`, `OnTaskSequence`, etc.

[har-transformer]: https://pypi.org/project/har-transformer
[Poetry]: https://github.com/sdispater/poetry

### Changed

  - Open-sourcing of this project in https://github.com/zalando-incubator.
  - `transformer.plugins.Plugin` is renamed
  `transformer.plugins.contracts.OnTaskSequence`.

[Unreleased]: https://github.com/zalando-incubator/transformer/compare/v1.0.2...HEAD
[1.0.2]: https://github.com/zalando-incubator/transformer/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/zalando-incubator/transformer/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/zalando-incubator/transformer/compare/f842c4163e037dc345eaf1992187f58126b7d909...v1.0.0

[@xinke2411]: https://github.com/xinke2411

[locust-dynamic-parameters]: https://docs.locust.io/en/stable/writing-a-locustfile.html#grouping-requests-to-urls-with-dynamic-parameters


update-version.py
#!/usr/bin/env python3

"""
update-version.py: Execute the first steps of Transformer's release process.

See https://transformer.readthedocs.io/en/latest/Versioning.html#release-process
for details.

Usage:
    update-version.py INCREMENT
    update-version.py --help
"""
import enum
import logging
import os
import subprocess
import tempfile
from pathlib import Path
from typing import Sequence

import dataclasses
import packaging.version
import pygments
import tomlkit
from dataclasses import dataclass
from docopt import docopt
from pygments.formatters.terminal import TerminalFormatter
from pygments.lexers.diff import DiffLexer


@dataclass
class Version:
    major: int
    minor: int
    patch: int

    @classmethod
    def parse(cls, s: str) -> "Version":
        v = packaging.version.parse(s)
        return Version(*v.release)

    def __str__(self) -> str:
        return ".".join((str(self.major), str(self.minor), str(self.patch)))


@dataclass
class Patch:
    target: os.PathLike
    lines: Sequence[str]

    def apply(self) -> bool:
        target = os.fspath(self.target)
        with tempfile.NamedTemporaryFile("a+") as f:
            f.write("\n".join(self.lines))
            f.write("\n")
            f.seek(0)
            result = subprocess.run(["patch", "-u", target], stdin=f)
        if result.returncode == 0:
            return True
        logging.error(f"failed to apply patch")
        logging.info(f"patch was {self!r}")
        return False

    def pretty(self) -> str:
        lines = list(self.lines)
        lines[0] += f" {self.target}"  # include target name in comments
        return pygments.highlight("\n".join(lines), DiffLexer(), TerminalFormatter())


def pyproject_patch(old_v: Version, new_v: Version) -> Patch:
    return Patch(
        target=Path("pyproject.toml"),
        lines=["@@ -3 +3 @@", f'-version = "{old_v}"', f'+version = "{new_v}"'],
    )


def sphinx_patch(old_v: Version, new_v: Version) -> Patch:
    old_short = f"{old_v.major}.{old_v.minor}"
    new_short = f"{new_v.major}.{new_v.minor}"
    return Patch(
        target=Path("docs", "conf.py"),
        lines=[
            "@@ -25,5 +25,5 @@",
            " # The short X.Y version",
            f'-version = "{old_short}"',
            f'+version = "{new_short}"',
            " # The full version, including alpha/beta/rc tags",
            f'-release = "{old_v}"',
            f'+release = "{new_v}"',
            " ",
        ],
    )


def changelog_patch(old_v: Version, new_v: Version) -> Patch:
    new_diff_url = (
        "https://github.com/zalando-incubator/transformer/compare/"
        f"v{old_v}...v{new_v}"
    )
    return Patch(
        target=Path("docs", "Changelog.rst"),
        lines=[
            "@@ -17,2 +17,11 @@",
            " ",
            f"+.. _v{new_v}:",
            "+",
            f"+v{new_v}",
            "+" + ("=" * (1 + len(str(new_v)))),
            "+",
            "+- Diff__.",
            "+",
            f"+__ {new_diff_url}",
            "+",
            f" .. _v{old_v}:",
        ],
    )


logging.basicConfig(
    level=logging.INFO, format="%(asctime)s\t%(levelname)s\t%(message)s"
)


class Increment(enum.IntEnum):
    MAJOR = 0
    MINOR = 1
    PATCH = 2


def increment_version(v: Version, incr: Increment) -> Version:
    values = list(dataclasses.astuple(v))
    values[incr.value] += 1
    for i in range(incr.value + 1, max(Increment).value + 1):
        values[i] = 0
    return Version(*values)


def pyproject_version(pyproject_path: os.PathLike) -> Version:
    pyproject_data = tomlkit.parse(Path(pyproject_path).read_text())
    return Version.parse(str(pyproject_data["tool"]["poetry"]["version"]))


def run():
    opts = docopt(__doc__)
    incr_str = opts["INCREMENT"].upper()
    try:
        incr: Increment = Increment[incr_str]
    except KeyError:
        logging.fatal(f"Invalid increment value {incr_str!r}")
        logging.info(f"Expecting one of {[i.name for i in Increment]}")
        raise  # PyCharm doesn't get that exit() prevents increments from not existing

    # Find the current version

    pyproject_path = Path("pyproject.toml")
    logging.info(f"Reading {pyproject_path} ...")
    old_version = pyproject_version(pyproject_path)
    logging.info(f"Current version: {old_version}")

    # Compute the new version, according to *incr*.
    new_version = increment_version(old_version, incr)
    logging.info(f"New version: {new_version} ({old_version} + {incr.name})")

    # Compute the corresponding patches.
    patches = [
        pyproject_patch(old_version, new_version),
        changelog_patch(old_version, new_version),
        sphinx_patch(old_version, new_version),
    ]

    # Show the patches, in case users want to reverse them later, and apply them.
    logging.info("Applying these patches:")
    for patch in patches:
        print(patch.pretty())
        patch.apply()
        print()


if __name__ == "__main__":
    run()
